src/backend/distributed/commands/local_multi_copy.c

```diff
@@ -0,0 +1,202 @@
+{
+	/* since we are doing a local copy, the following statements should use local execution to see the changes */
+	TransactionAccessedLocalPlacement = true;
+
+	bool isBinaryCopy = localCopyOutState->binary;
+	if (ShouldAddBinaryHeaders(localCopyOutState->fe_msgbuf, isBinaryCopy))
+	{
+		AppendCopyBinaryHeaders(localCopyOutState);
+	}
+
+	AddSlotToBuffer(slot, copyDest, localCopyOutState);
+
+	if (ShouldSendCopyNow(localCopyOutState->fe_msgbuf))
+	{
+		if (isBinaryCopy)
+		{
+			AppendCopyBinaryFooters(localCopyOutState);
+		}
+		bool isEndOfCopy = false;
+		DoLocalCopy(localCopyOutState->fe_msgbuf, copyDest->distributedRelationId,
+					shardId,
+					copyDest->copyStatement, isEndOfCopy);
+	}
+}
+
+
+/*
+ * FinishLocalCopyToShard finishes local copy for the given shard with the shard id.
+ */
+void
+FinishLocalCopyToShard(CitusCopyDestReceiver *copyDest, int64 shardId,
+					   CopyOutState localCopyOutState)
+{
+	bool isBinaryCopy = localCopyOutState->binary;
+	if (isBinaryCopy)
+	{
+		AppendCopyBinaryFooters(localCopyOutState);
+	}
+	bool isEndOfCopy = true;
+	DoLocalCopy(localCopyOutState->fe_msgbuf, copyDest->distributedRelationId, shardId,
+				copyDest->copyStatement, isEndOfCopy);
+}
+
+
+/*
+ * AddSlotToBuffer serializes the given slot and adds it to the buffer in localCopyOutState.
+ */
+static void
+AddSlotToBuffer(TupleTableSlot *slot, CitusCopyDestReceiver *copyDest, CopyOutState
+				localCopyOutState)
+{
+	Datum *columnValues = slot->tts_values;
+	bool *columnNulls = slot->tts_isnull;
+	FmgrInfo *columnOutputFunctions = copyDest->columnOutputFunctions;
+	CopyCoercionData *columnCoercionPaths = copyDest->columnCoercionPaths;
+
+	AppendCopyRowData(columnValues, columnNulls, copyDest->tupleDescriptor,
+					  localCopyOutState, columnOutputFunctions,
+					  columnCoercionPaths);
+}
+
+
+/*
+ * ShouldSendCopyNow returns true if the given buffer size exceeds the
+ * local copy buffer size threshold.
+ */
+static bool
+ShouldSendCopyNow(StringInfo buffer)
+{
+	return buffer->len > LOCAL_COPY_FLUSH_THRESHOLD;
+}
+
+
+/*
+ * DoLocalCopy finds the shard table from the distributed relation id, and copies the given
+ * buffer into the shard.
+ */
+static void
+DoLocalCopy(StringInfo buffer, Oid relationId, int64 shardId, CopyStmt *copyStatement,
+			bool isEndOfCopy)
+{
+	LocalCopyBuffer = buffer;
+
+	Oid shardOid = GetShardLocalTableOid(relationId, shardId);
+	Relation shard = heap_open(shardOid, RowExclusiveLock);
+	ParseState *pState = make_parsestate(NULL);
+
+	/* p_rtable of pState is set so that we can check constraints. */
+	pState->p_rtable = CreateRangeTable(shard, ACL_INSERT);
+
+	CopyState cstate = BeginCopyFrom(pState, shard, NULL, false,
+									 ReadFromLocalBufferCallback,
+									 copyStatement->attlist, copyStatement->options);
+	CopyFrom(cstate);
+	EndCopyFrom(cstate);
+
+	heap_close(shard, NoLock);
+	free_parsestate(pState);
+	FreeStringInfo(buffer);
+	if (!isEndOfCopy)
+	{
+		buffer = makeStringInfo();
```

Comment:
could use resetStringInfo

Reply:
We call FreeStringInfo, which already calls resetStringInfo:

```
void
FreeStringInfo(StringInfo stringInfo)
{
	resetStringInfo(stringInfo);

	pfree(stringInfo->data);
	pfree(stringInfo);
}
```

I thought of using `FreeStringInfo` so that we will be freeing the data as well, which could result in less memory usage at times.

Reply:
It doesn't make much sense now. `buffer` is a local variable so the `makeStringInfo()` result goes nowhere, and don't we need fe_msgbuf for the next cycle?

This probably requires some tests that exceed LOCAL_COPY_FLUSH_THRESHOLD.

Reply:
maybe I am missing something but we call DoLocalCopy with fe_msgbuf, and StringBuffer is a pointer type, so before we go to the next cycle I thought of freeing both the internal buffer that is kept in fe_msgbuf, but then we freed fe_msgbuf and we should have fe_msgbuf for the next cycle, therefore there is MakeStringInfo if it is not end of the copy.

I have debugged it and verified that the pointer that we have in this method is the same pointer as fe_msgbuf.

With benchmarking I have tested the case where we exceed LOCAL_COPY_FLUSH_THRESHOLD but I will add a test for it

Reply:
`buffer` is a local variable that's declared here:

```c
DoLocalCopy(StringInfo buffer, Oid relationId, int64 shardId, CopyStmt *copyStatement,
			bool isEndOfCopy)
```

Assigning a local variable at the end of the function does nothing.

`FreeStringInfo` calls pfree on the StringInfo pointed to by `fe_msgbuf`. Using it after this function will likely cause a segfault.